File: 42. Trapping Rain Water/main.c
int trap(int *height, int heightSize) {
    int left = 0;
    int right = heightSize - 1;
    int leftMax = 0;
    int rightMax = 0;
    int total = 0;

    while (left <= right) {
        if (height[left] <= height[right]) {
            if (height[left] <= leftMax) {
                total += leftMax - height[left];
            } else {
                leftMax = height[left];
            }
            left++;
        } else {
            if (height[right] <= rightMax) {
                total += rightMax - height[right];
            } else {
                rightMax = height[right];
            }
            right--;
        }
    }

    return total;
}

static int max(int a, int b) {
    return (a > b) ? a : b;
}

static int min(int a, int b) {
    return (a < b) ? a : b;
}

int trap(int *height, int heightSize) {
    int prefixMax[heightSize];
    prefixMax[0] = 0;

    for (int i = 1; i < heightSize; i++) {
        prefixMax[i] = max(prefixMax[i - 1], height[i - 1]);
    }

    int suffixMax[heightSize];
    suffixMax[heightSize - 1] = 0;

    for (int i = heightSize - 2; i >= 0; i--) {
        suffixMax[i] = max(suffixMax[i + 1], height[i + 1]);
    }

    int total = 0;

    for (int i = 0; i < heightSize; i++) {
        int top = min(prefixMax[i], suffixMax[i]);
        total += (top >= height[i]) ? (top - height[i]) : 0;
    }

    return total;
}

File: 42. Trapping Rain Water/main.rs
struct Solution;

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let size = height.len();
        let mut left = 0;
        let mut right = size as i32 - 1;
        let mut left_max = 0;
        let mut right_max = 0;
        let mut total = 0;

        while left <= right {
            if height[left as usize] <= height[right as usize] {
                if height[left as usize] <= left_max {
                    total += left_max - height[left as usize];
                } else {
                    left_max = height[left as usize];
                }
                left += 1;
            } else {
                if height[right as usize] <= right_max {
                    total += right_max - height[right as usize];
                } else {
                    right_max = height[right as usize];
                }
                right -= 1;
            }
        }

        total
    }
}

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let size = height.len();
        let mut prefix_max = vec![0; size];

        for i in 1..size {
            prefix_max[i] = prefix_max[i - 1].max(height[i - 1]);
        }

        let mut suffix_max = vec![0; size];

        for i in (0..size - 1).rev() {
            suffix_max[i] = suffix_max[i + 1].max(height[i + 1]);
        }

        let mut total = 0;

        for i in 0..size {
            let top = prefix_max[i].min(suffix_max[i]);
            if top >= height[i] {
                total += top - height[i];
            }
        }

        total
    }
}
